File: javascript/foundry-local.mjs
import { OpenAI } from "openai";
import { FoundryLocalManager } from "foundry-local-sdk";

// By using an alias, the most suitable model variant will be
// downloaded for your end-user's device hardware.
const alias = "phi-3.5-mini";

// Step 1: Create a FoundryLocalManager and start the service
console.log("Starting Foundry Local service...");
FoundryLocalManager.create({ appName: "FoundryLocalWorkshop" });
const manager = FoundryLocalManager.instance;
await manager.startWebService();

// Step 2: Get the model from the catalog
const catalog = manager.catalog;
const model = await catalog.getModel(alias);

// Step 3: Download model if needed, or skip
if (model.isCached) {
  console.log(`Model already downloaded: ${alias}`);
} else {
  console.log(
    `Downloading model: ${alias} (this may take several minutes)...`
  );
  await model.download();
  console.log(`Download complete: ${alias}`);
}

// Step 4: Load the model into memory
console.log(`Loading model: ${alias}...`);
await model.load();
console.log(`Model loaded: ${model.id}`);

// Configure the OpenAI client to use the local Foundry service
const client = new OpenAI({
  baseURL: manager.urls[0] + "/v1",
  apiKey: "foundry-local", // API key is not required for local usage
});

// Generate a streaming chat completion
const stream = await client.chat.completions.create({
  model: model.id,
  messages: [{ role: "user", content: "What is the golden ratio?" }],
  stream: true,
});

// Print the streaming response
for await (const chunk of stream) {
  if (chunk.choices[0]?.delta?.content) {
    process.stdout.write(chunk.choices[0].delta.content);
  }
}
console.log(); // newline at end

// Cleanup: unload the model to release resources
await model.unload();
await manager.stopWebService();

File: python/foundry-local.py

import openai
from foundry_local import FoundryLocalManager

# By using an alias, the most suitable model variant will be
# downloaded for your end-user's device hardware.
alias = "phi-3.5-mini"

# Step 1: Create a FoundryLocalManager and start the service
print("Starting Foundry Local service...")
manager = FoundryLocalManager()
manager.start_service()

# Step 2: Check if the model is already downloaded
cached = manager.list_cached_models()
catalog_info = manager.get_model_info(alias)
is_cached = any(m.id == catalog_info.id for m in cached) if catalog_info else False

if is_cached:
    print(f"Model already downloaded: {alias}")
else:
    print(f"Downloading model: {alias} (this may take several minutes)...")
    manager.download_model(alias)
    print(f"Download complete: {alias}")

# Step 3: Load the model into memory
print(f"Loading model: {alias}...")
manager.load_model(alias)
print(f"Cached models: {[m.id for m in cached]}")

# Configure the OpenAI client to use the local Foundry service.
# Foundry Local assigns a dynamic port — always use manager.endpoint.
client = openai.OpenAI(
    base_url=manager.endpoint,
    api_key=manager.api_key  # API key is not required for local usage
)

# Generate a streaming chat completion
stream = client.chat.completions.create(
    model=manager.get_model_info(alias).id,
    messages=[{"role": "user", "content": "What is the golden ratio?"}],
    stream=True,
)

# Print the streaming response
for chunk in stream:
    if chunk.choices[0].delta.content is not None:
        print(chunk.choices[0].delta.content, end="", flush=True)
print()

# Cleanup: unload the model to release resources
manager.unload_model(alias)  # newline at end